File: benchmarks/vm/11_union_match_loop.rb
Value = Struct.new(:tag, :payload)

def make_value(i)
  case i % 3
  when 0
    Value.new(:a, i)
  when 1
    Value.new(:b, i * 2)
  else
    Value.new(:c, nil)
  end
end

t0 = Process.clock_gettime(Process::CLOCK_MONOTONIC)
total = 0
100_000.times do |i|
  v = make_value(i)
  case v.tag
  when :a
    total += v.payload
  when :b
    total -= v.payload
  else
    total += 1
  end
end
total_ms = (Process.clock_gettime(Process::CLOCK_MONOTONIC) - t0) * 1000
puts "total = #{total}"
puts format("BENCH_RESULT: %.3f ms", total_ms)

File: benchmarks/vm/11_union_match_loop.lua
local function make_value(i)
    local r = i % 3
    if r == 0 then
        return { tag = "a", payload = i }
    elseif r == 1 then
        return { tag = "b", payload = i * 2 }
    end
    return { tag = "c" }
end

local t0 = os.clock()
local total = 0
for i = 0, 99999 do
    local v = make_value(i)
    if v.tag == "a" then
        total = total + v.payload
    elseif v.tag == "b" then
        total = total - v.payload
    else
        total = total + 1
    end
end
local total_ms = (os.clock() - t0) * 1000
print("total = " .. tostring(total))
print(string.format("BENCH_RESULT: %.3f ms", total_ms))
